### PR TITLE
board/Nucleo-f303: initial support for the Nucleo-f303

### PR DESCRIPTION
**Pull request by katezilla**
Add support for Nucleo-f303 (STM32f303)

### CODE REVIEW

**Comment by BytesGalore** [comment]
Congrats for your first PR :)

File: boards/nucleo-f303/Makefile.include [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+PORT_DARWIN ?= $(shell ls -1 /dev/tty.usbmodem* | head -n 1)
+
+# setup serial terminal
+include $(RIOTBOARD)/Makefile.include.serial
+
+# this board uses openocd
+include $(RIOTBOARD)/Makefile.include.openocd
+
+# include cortex defaults
+include $(RIOTBOARD)/Makefile.include.cortexm_common
```
Comment (BytesGalore) [inline boards/nucleo-f303/Makefile.include]: please add a newline here

File: boards/nucleo-f303/include/board.h [hunk 2]
```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (C) 2015  Freie Universität Berlin
```
Comment (BytesGalore) [inline boards/nucleo-f303/include/board.h]: please adjust the license header

File: boards/nucleo-f303/include/board.h [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+/**
+ * @defgroup    boards_nucleo-f303 Nucleo-F303
+ * @ingroup     boards
+ * @brief       Board specific files for the nucleo-f303 board
+ * @{
+ *
+ * @file
+ * @brief       Board specific definitions for the nucleo-f303 board
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
```
Comment (BytesGalore) [inline boards/nucleo-f303/include/board.h]: Please add yourself here

Comment (BytesGalore) [inline boards/nucleo-f303/include/board.h]: Or is this header just a copy?

File: boards/nucleo-f303/include/periph_conf.h [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+#define TIMER_IRQ_PRIO      1
+
+/* Timer 0 configuration */
+#define TIMER_0_DEV         TIM2
+#define TIMER_0_CHANNELS    4
+#define TIMER_0_PRESCALER   (71U)
+#define TIMER_0_MAX_VALUE   (0xffffffff)
+#define TIMER_0_CLKEN()     (RCC->APB1ENR |= RCC_APB1ENR_TIM2EN)
+#define TIMER_0_IRQ_CHAN    TIM2_IRQn
+#define TIMER_0_ISR         isr_tim2 // ka wo her
```
Comment (BytesGalore) [inline boards/nucleo-f303/include/periph_conf.h]: Please remove the comment

File: cpu/stm32f3/ldscripts/stm32f303re.ld [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+MEMORY
+{
+    rom (rx)    : ORIGIN = 0x08000000, LENGTH = 512K
+    ram (rwx)   : ORIGIN = 0x20000000, LENGTH = 64K
+    ccmram (rwx): ORIGIN = 0x10000000, LENGTH = 16K
+}
+
+INCLUDE cortexm_base.ld
```
Comment (BytesGalore) [inline cpu/stm32f3/ldscripts/stm32f303re.ld]: and please add a newline here :)

File: boards/nucleo-f303/include/board.h [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+#include "periph_conf.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @name Define the nominal CPU core clock in this board
+ */
+#define F_CPU               CLOCK_CORECLOCK
```
Comment (BytesGalore) [inline boards/nucleo-f303/include/board.h]: please add brackets (also for `UART_0` and `TIMER_0` below)

Comment (BytesGalore) [inline boards/nucleo-f303/include/board.h]: ok forget this, it seems we never used brackets for these cases.

**Comment by BytesGalore** [comment]
alright, the white-space warnings are history :+1: 
I think you can squash your commits now, then lets see what @PeterKietzmann says.

File: boards/nucleo-f303/include/periph_conf.h [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @name Clock system configuration
+ * @{
+ **/
+#define CLOCK_HSE         (8000000U)          /* external oscillator */
+#define CLOCK_CORECLOCK   (72000000U)         /* desired core clock frequency */
```
Comment (haukepetersen) [inline boards/nucleo-f303/include/periph_conf.h]: the indention seems to be of by a few spaces here.

**Comment by haukepetersen** [comment]
Nice one! Except some minor style issues I think the PR looks good. Codewise you get my ack once the comments above are addressed. I don't have the board here, @BytesGalore and @PeterKietzmann, will you test it?

Small thing when squashing: try to give your commits some good names. The convention used by most is something like `MAIN_MODULE/MODULE: this is what I did`, e.g. `boards: added support for the nucleo-f303` or something like this...

File: boards/nucleo-f303/include/periph_conf.h (hunk 7 above)
Comment (katezilla) [inline boards/nucleo-f303/include/periph_conf.h]: less spaces or more for it to be on same collum as the other comments?

Comment (haukepetersen) [inline boards/nucleo-f303/include/periph_conf.h]: I would put it to the same indention as the lines below.

File: boards/nucleo-f303/include/periph_conf.h [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+ *
+ * @file
+ * @brief       Peripheral MCU configuration for the nucleo-f303 board
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Katja Kirstein <katja.kirstein@haw-hamburg.de>
+ */
+
+#ifndef __PERIPH_CONF_H
+#define __PERIPH_CONF_H
```
Comment (gebart) [inline boards/nucleo-f303/include/periph_conf.h]: Change this include guard to `PERIPH_CONF_H_` instead. Double underscores are reserved.

File: boards/nucleo-f303/include/periph_conf.h [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+#define GPIO_2_EXTI_CFG1()  (SYSCFG->EXTICR[0] &= ~(SYSCFG_EXTICR1_EXTI2))
+#define GPIO_2_EXTI_CFG2()  (SYSCFG->EXTICR[0] |= SYSCFG_EXTICR1_EXTI2_PD)
+#define GPIO_2_IRQ          EXTI2_TSC_IRQn
+/** @} */
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* __PERIPH_CONF_H */
```
Comment (gebart) [inline boards/nucleo-f303/include/periph_conf.h]: missing newline at end of file. Also remember to update the comment when you change the include guard name.

File: boards/nucleo-f303/board.c [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+/**
+ * @brief Initialize the boards on-board LEDs
+ *
+ * The LED initialization is hard-coded in this function.
+ * As the LED is soldered onto the board it is fixed to
+ * its CPU pins.
+ *
+ * The green LED is connected to pin PA5
```
Comment (PeterKietzmann) [inline boards/nucleo-f303/board.c]: Is this the case or just copy-paste?

File: boards/nucleo-f303/board.c [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    /* set output type to push-pull */
+    LED_GREEN_PORT->OTYPER &= ~(0x00000020);
+    /* configure pin as general output */
+    LED_GREEN_PORT->MODER &= ~(0x00000c00);
+    LED_GREEN_PORT->MODER |= 0x00000400;
+    /* disable pull resistors */
+    LED_GREEN_PORT->PUPDR &= ~(0x00000c00);
+    /* turn all LEDs off */
+    LED_GREEN_PORT->BRR = 0x00c0;
+}
```
Comment (PeterKietzmann) [inline boards/nucleo-f303/board.c]: Did you test the LED?

File: boards/nucleo-f303/include/board.h [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+#define LED_RED_OFF
+#define LED_RED_TOGGLE
+
+#define LED_GREEN_ON        (LED_GREEN_PORT->BSRRL = (1 << LED_GREEN_PIN))
+#define LED_GREEN_OFF       (LED_GREEN_PORT->BSRRH = (1 << LED_GREEN_PIN))
+#define LED_GREEN_TOGGLE    (LED_GREEN_PORT->ODR ^= (1 << LED_GREEN_PIN))
+
+#define LED_ORANGE_ON
+#define LED_ORANGE_OFF
+#define LED_ORANGE_TOGGLE
```
Comment (PeterKietzmann) [inline boards/nucleo-f303/include/board.h]: See above! Also: Aren't there more LEDs on this board?

**Comment by PeterKietzmann** [comment]
@katezilla, looks good :-) ! Congrats for your first PR. Please address the minor style comments and just ask if something is unclear.

File: boards/nucleo-f303/board.c (hunk 10 above)
Comment (haukepetersen) [inline boards/nucleo-f303/board.c]: I think the LED is actually connected to the pin PA5 on every single nucleo board.

File: boards/nucleo-f303/include/board.h (hunk 12 above)
Comment (katezilla) [inline boards/nucleo-f303/include/board.h]: there are three LEDs but only one is for the user and i tested that one.

**Comment by PeterKietzmann** [comment]
@katezilla thanks for the adaptions. Could you also address the newline-comments please? Otherwise we're not far from merging this PR :-)

**Comment by PeterKietzmann** [comment]
Sorry, now I see you already addressed the newline comments. Could you please add an I2C configuration to the periph_conf as a last step :-) ?

**Comment by PeterKietzmann** [comment]
@katezilla I tested I2C. For some reasons I2C_0 works and I2C_1 does not. Any ideas? Anyway, I think you can already squash your commits to one. I think your first commit message is fine.

**Comment by PeterKietzmann** [comment]
Does anyone have a stm32f3discovery board and can test the I2C_1 driver?

**Comment by PeterKietzmann** [comment]
@katezilla you defined pins `PD12...15` for the `PWM_1_DEV`. Where do I access them on the board. Is [this](https://developer.mbed.org/media/uploads/bcostm/nucleo64_revc_f303re_mbed_pinout_v3_morpho.png) schematic correct?

**Comment by PeterKietzmann** [comment]
I did the following measurement with the settings:
- frequency = 1000 Hz
- duty-cycle = 10us
bildschirmfoto vom 2015-06-12 09_24_21
It appears that the period time and the duty cycle are half as long as expected. Seems like a clock configuration thing. @haukepetersen do you know where the bug is **or** is it maybe in the driver? Can someone confirm with an stm32f3discovery

**Comment by haukepetersen** [comment]
Just did. It's a bug in the frequency calculation of the f3 pwm driver -> unrelated to this PR. Lets just merge this board as is, but please open an issue so we can fix it afterwards for all f3 boards!

**Comment by PeterKietzmann** [comment]
ACK! I'll merge when Travis is green. Jippie!

**Comment by haukepetersen** [comment]
removed squashing label and restarted Travis

**Comment by PeterKietzmann** [comment]
The cortex-m4 build fails on the ported nucleo-f303 board for the both tests `driver_at86rf2xx` and `driver_isl29125`. Can anyone imagine why?

**Comment by gebart** [comment]
Just a thought: the build errors could be caused by missing gpio
definitions or similar. Check periph_conf.h

**Comment by PeterKietzmann** [comment]
That would make sense cause the one test requires `GPIO_5` and the other `GPIO_0 ... GPIO_4`. In this PR we just introduced the GPIOs `GPIO_0 ... GPIO_2`. I'd prefer to merge this PR quickly without many (big) extensions. Do we have some kind of application blacklist? I don't know...

**Comment by gebart** [comment]
There is a blacklist feature in the application Makefile. I don't remember
the exact name, something like BLACKLIST_BOARDS, look around in the
Makefiles of other test applications for it.

**Comment by haukepetersen** [comment]
Actually, the GPIO issue should be obsolete as #3095 is merged. To adapt this PR simply rebase this PR on master and remove the complete GPIO section from the `periph_conf.h`. Then Travis should run through just fine.

**Comment by PeterKietzmann** [comment]
@haukepetersen do I see it correctly that just the `GPIO_NUMUF` needs to be removed for adaptation to #3095?

**Comment by PeterKietzmann** [comment]
ACK and go. Will do a PR to remove the superfluous macro

**Comment by PeterKietzmann** [comment]
Congrats for your first PR @katezilla!

**Comment by katezilla** [comment]
thx :)

**Pull request merged**